lib/watir-webdriver/locators/element/locator.rb

```diff
@@ -0,0 +1,247 @@
+module Watir
+  class Element
+    class Locator
+      attr_reader :selector_builder
+      attr_reader :element_validator
+
+      WD_FINDERS = [
+        :class,
+        :class_name,
+        :css,
+        :id,
+        :link,
+        :link_text,
+        :name,
+        :partial_link_text,
+        :tag_name,
+        :xpath
+      ]
+
+      # Regular expressions that can be reliably converted to xpath `contains`
+      # expressions in order to optimize the .
+      CONVERTABLE_REGEXP = %r{
+        \A
+          ([^\[\]\\^$.|?*+()]*) # leading literal characters
+          [^|]*?                # do not try to convert expressions with alternates
+          ([^\[\]\\^$.|?*+()]*) # trailing literal characters
+        \z
+      }x
+
+      def initialize(parent, selector, selector_builder, element_validator)
+        @parent = parent # either element or browser
+        @selector = selector.dup
+        @selector_builder = selector_builder
+        @element_validator = element_validator
+      end
+
+      def locate
+        e = by_id and return e # short-circuit if :id is given
+
+        if @selector.size == 1
+          element = find_first_by_one
+        else
+          element = find_first_by_multiple
+        end
+
+        # This actually only applies when finding by xpath/css - browser.text_field(:xpath, "//input[@type='radio']")
+        # We don't need to validate the element if we built the xpath ourselves.
+        # It is also used to alter behavior of methods locating more than one type of element
+        # (e.g. text_field locates both input and textarea)
+        element_validator.validate(element, @selector) if element
```

Comment:
I think we can restrict this more to just the elements that need it. Should get a small performance improvement on every location call.